numpy/lib/index_tricks.py

```diff
@@ -238,45 +238,41 @@ class AxisConcatenator(object):
 
     """
 
-    def _retval(self, res):
-        if self.matrix:
-            oldndim = res.ndim
-            res = makemat(res)
-            if oldndim == 1 and self.col:
-                res = res.T
-        self.axis = self._axis
-        self.matrix = self._matrix
-        self.col = 0
-        return res
```

Comment:
Problem was that this cleanup wouldn't run if there was an error. Better to not break things to the point that they need cleanup in the first place.

Also, this means that `np.r_` wasn't threadsafe. Hooray.